File: pytango/component/Component_3.cpp
static const char *RcsId = "$Id:  $";

#include <Component_3.h>
#include <Component_3Class.h>


namespace Component_3_ns
{
void PA_callback::push_event(Tango::DataReadyEventData *event)
{
	 (static_cast<Component_3 *>(device))->pa(event);
	
}


Component_3::Component_3(Tango::DeviceClass *cl, string &s)
 : TANGO_BASE_CLASS(cl, s.c_str())
{
	init_device();
	
}

Component_3::Component_3(Tango::DeviceClass *cl, const char *s)
 : TANGO_BASE_CLASS(cl, s)
{
	init_device();
	
}

Component_3::Component_3(Tango::DeviceClass *cl, const char *s, const char *d)
 : TANGO_BASE_CLASS(cl, s, d)
{
	init_device();

}


void Component_3::delete_device()
{
	DEBUG_STREAM << "Component_3::delete_device() " << device_name << endl;
}


void Component_3::init_device()
{
	DEBUG_STREAM << "Component_3::init_device() create device " << device_name << endl;

	set_state(Tango::ON);
	attr_c_state = false;
	c1_dev = new Tango::DeviceProxy("sys/component_1/C1");
	c2_dev = new Tango::DeviceProxy("sys/component_2/C2");
    pa_callback = new PA_callback(this);
}


void Component_3::always_executed_hook()
{
	INFO_STREAM << "Component_3::always_executed_hook()  " << device_name << endl;
	c1_dev->subscribe_event("PA", Tango::DATA_READY_EVENT, static_cast<Tango::CallBack *>(pa_callback));
}

//--------------------------------------------------------
/**
 *	Method      : Component_3::read_attr_hardware()
 *	Description : Hardware acquisition for attributes
 */
//--------------------------------------------------------
void Component_3::read_attr_hardware(TANGO_UNUSED(vector<long> &attr_list))
{
	DEBUG_STREAM << "Component_3::read_attr_hardware(vector<long> &attr_list) entering... " << endl;

}


void Component_3::add_dynamic_attributes()
{

}


void Component_3::toggle()
{
	DEBUG_STREAM << "Component_3::toggle()  - " << device_name << endl;
	attr_c_state = not attr_c_state;
}

void Component_3::re()
{
	DEBUG_STREAM << "Component_3::RE()  - " << device_name << endl;
	if(attr_c_state == false)
	{
		toggle();
	}
}

void Component_3::rl()
{
	DEBUG_STREAM << "Component_3::RL()  - " << device_name << endl;
	if(attr_c_state == true)
	{
		toggle();
		c2_dev->command_inout("dr");
	}
}

void Component_3::pa(Tango::DataReadyEventData *event)
{
	DEBUG_STREAM << "Component_3::pa()  - " << device_name << endl;
	attr_c_state = false;
}

} //	namespace

File: pytango/component/component_3.py
# This code is part of the checkmate project.
# Copyright (C) 2013-2016 The checkmate project contributors
#
# This program is free software under the terms of the GNU GPL, either
# version 3 of the License, or (at your option) any later version.

import sys

import PyTango


class Component_3(PyTango.Device_4Impl):
    def __init__(self, _class, name):
        super(Component_3, self).__init__(_class, name)
        Component_3.init_device(self)
        self.received_exchange = []

    def init_device(self):
        self.get_device_properties(self.get_device_class())
        self.set_state(PyTango.DevState.ON)
        self.attr_c_state = False
        self.c1_dev = PyTango.DeviceProxy('sys/component_1/C1')
        self.c2_dev = PyTango.DeviceProxy('sys/component_2/C2')

    def toggle(self):
        self.attr_c_state = not self.attr_c_state

    def RE(self):
        if self.attr_c_state == False:
            self.toggle()

    def RL(self):
        if self.attr_c_state == True:
            self.toggle()
            self.c2_dev.DR()

    def PA(self):
        if self.attr_c_state == False:
            pass


class C3Interface(PyTango.DeviceClass):
    def dyn_attr(self, dev_list):
        """"""
        for dev in dev_list:
            try:
                dev.initialize_dynamic_attributes()
            except:
                import traceback
                dev.warn_stream("Failed to initialize dynamic attributes")
                dev.debug_stream("Details: " + traceback.format_exc())
 

    cmd_list = {'RE': [[PyTango.DevVoid], [PyTango.DevVoid]],
                'RL': [[PyTango.DevVoid], [PyTango.DevVoid]],
                'PA': [[PyTango.DevVoid], [PyTango.DevVoid]]
               }
    attr_list = {
                }


if __name__ == '__main__':
    py = PyTango.Util(sys.argv)
    py.add_class(C3Interface, Component_3, 'Component_3')
    U = PyTango.Util.instance()
    U.server_init()
    U.server_run()

